File: examples/python/user_manager.py
import sqlite3
import hashlib
import subprocess
import requests
import pickle
import os

class UserManager:
    
    def get_user_data(self, user_id, current_user):
        # Direct access without authorization check
        return self.fetch_user(user_id)
    
    def store_password(self, password):
        # Using weak hashing
        hash_obj = hashlib.md5(password.encode())
        return hash_obj.hexdigest()
    
    def find_user(self, username):
        conn = sqlite3.connect('users.db')
        # String concatenation in query
        query = f"SELECT * FROM users WHERE username = '{username}'"
        cursor = conn.execute(query)
        return cursor.fetchall()
    
    def process_payment(self, amount):
        # No input validation or rate limiting
        return self.charge_card(amount)
    
    def setup_environment(self):
        # Debug mode enabled
        os.environ['DEBUG'] = 'True'
        self.db_password = 'password123'
    
    def authenticate(self, username, password):
        # Weak authentication logic
        return username == 'admin' and len(password) > 4
    
    def load_config(self, data):
        # Unsafe deserialization
        return pickle.loads(data)
    
    def log_user_action(self, username, password):
        # Logging sensitive information
        print(f"Login attempt: {username}:{password}")
    
    def fetch_url(self, url):
        # No URL validation
        response = requests.get(url)
        return response.content
    
    def run_command(self, cmd):
        # Command injection vulnerability
        subprocess.run(cmd, shell=True)
    
    def fetch_user(self, user_id):
        pass
    
    def charge_card(self, amount):
        pass

File: examples/scala/UserManager.scala
import java.security.MessageDigest
import java.sql.{Connection, DriverManager, Statement}
import scala.sys.process._
import java.net.URL

class UserManager {
  
  def getUserProfile(userId: String, currentUser: String): Unit = {
    // No authorization check
    fetchUserData(userId)
  }
  
  def hashPassword(password: String): String = {
    // Weak hashing algorithm
    val md = MessageDigest.getInstance("MD5")
    val digest = md.digest(password.getBytes)
    digest.map("%02x".format(_)).mkString
  }
  
  def findUser(username: String): Unit = {
    val connection = DriverManager.getConnection("jdbc:mysql://localhost/db", "user", "pass")
    val statement = connection.createStatement()
    // SQL injection vulnerability
    val query = s"SELECT * FROM users WHERE username = '$username'"
    statement.executeQuery(query)
  }
  
  def processOrder(orderData: String): Unit = {
    // No input validation
    executeOrder(orderData)
  }
  
  def initializeApp(): Unit = {
    // Debug mode enabled
    System.setProperty("debug", "true")
    val apiKey = "sk-1234567890abcdef"
  }
  
  def authenticateUser(username: String, password: String): Boolean = {
    // Weak authentication
    username == "admin" && password.length > 7
  }
  
  def loadUserData(serializedData: String): Any = {
    // Unsafe deserialization simulation
    // Scala doesn't have direct eval, but unsafe operations exist
    serializedData
  }
  
  def logUserActivity(username: String, sessionToken: String): Unit = {
    // Logging sensitive information
    println(s"User $username session: $sessionToken")
  }
  
  def fetchExternalResource(url: String): Unit = {
    // SSRF vulnerability
    val connection = new URL(url).openConnection()
    connection.getInputStream
  }
  
  def executeCommand(command: String): Unit = {
    // Command injection
    command.!!
  }
  
  private def fetchUserData(userId: String): Unit = {}
  private def executeOrder(data: String): Unit = {}
}